File: HW11-2DDarcy-Stokes-Steady/Python/darcy_stokes_validation_compaction_colum.py
# Darcy-Stokes solver
# author: Mohammad Afzal Shadab
# email: mashadab@utexas.edu
# date: 04/09/2023

import sys
sys.path.insert(1, '../../HW1-Numerics/Python/')
sys.path.insert(1, '../../HW2-BC_LBVP/Python/')
sys.path.insert(1, '../../HW3-Hetero-Fluxes-NeuBC/Python/')
sys.path.insert(1, '../../HW4-1DMelt-Migration/Python/')
sys.path.insert(1, '../../HW6-2D_operators/Python/')
sys.path.insert(1, '../../HW8-2DStokes/Python/')
sys.path.insert(1, '../../HW9-Streamlines/Python/')

# import Python libraries
from scipy.sparse import bmat,csr_matrix,spdiags
import matplotlib.pyplot as plt

# import personal libraries
from classfun import *  #importing the classes and relevant functions
from build_stokes_grid_fun import build_stokes_grid
from build_stokes_ops_Darcy_Stokes_fun import build_stokes_ops_Darcy_Stokes
from build_bndfun_optimized import build_bnd
from solve_lbvpfun_optimized import solve_lbvp
from quiver_plot import quiver_plot
from comp_streamfun import comp_streamfun
from plot_streamlines import plot_streamlines

def build_Zd(G,phi,m,mu,Grid): #building zeta^*_phi at cell centers
    Zd = (G / (phi**m) - 2/3) * mu * (1-phi);
    Zd = spdiags(Zd.T,0,Grid.N,Grid.N)
    return Zd

def build_Kd(k0,n,phi,mu_f,Grid,Mp): #building Kd at cell faces
    Kd = Mp @ ( k0 * phi**n / mu_f)
    Kd = spdiags(Kd.T,0,Grid.Nf,Grid.Nf)
    return Kd 

def build_RHS(phi,Kd,Grid,Mp,Dp,rho_f,rho_s,Gamma,grav):
    #fv at cell faces
    fv =-grav*( Mp @ ((rho_f - rho_s) *(1-phi))) * np.vstack([np.zeros((Grid.Nfx,1)),np.ones((Grid.Nfy,1))])
    
    #fh at cell centers
    fh =-(rho_f - rho_s)/(rho_f * rho_s) * Gamma * np.ones((Grid.N,1))
    
    F =  np.vstack([fv,fh])
    
    return F

#problem parameter
mu_max = 1e14 #Maximum solid viscosity [Pa.s]
phi_min= 1e-2; phi_max= 0.7 #Minimum and Maximum solid porosities [-]
G      = 1.0 #Coefficient in the bulk viscosity formulation [-]
k0     = 5.6e-11 #Relative permeability [m^2] {Meyer and Hewitt, 2017}
mu_f   = 1e-3#Viscosity of the fluid [Pa.s]
m      = 1   #Power law coefficient in compaction viscosity = (G/phi^m * mu_s) [-]
n      = 2   #Power law coefficient in porosity permeability relationship k = k0*phi^n [-]
rho_s  = 917 #Density of solid [kg/m^3]
rho_f  = 1e3 #Density of fluid [kg/m^3]
Gamma  = 0   #Rate of melting [kg/m^3-s]
grav   = 9.81#Acceleration due to gravity []m/s^2]
Delta_rho = rho_f - rho_s #Density difference between the two phases [kg/m^3]
vt     = 1e-3#Tangential velocity [m/s]

#Compaction length
phic = phi_min
delta0 = np.sqrt(k0*phic**n*mu_max/(phic**m*mu_f))
Kc = k0*Delta_rho*grav*phic**n/mu_f

#building grid
Gridp.xmin = 0.0*delta0 ; Gridp.xmax = 1*delta0; Gridp.Nx   = 5
Gridp.ymin = 0.0*delta0 ; Gridp.ymax = 1*delta0; Gridp.Ny   = 100
Gridp.geom = 'cartesian'
Grid = build_stokes_grid(Gridp)
[Xc,Yc] = np.meshgrid(Gridp.xc,Gridp.yc)
Xc_col = np.reshape(Xc.T,(Gridp.N,-1)); Yc_col = np.reshape(Yc.T,(Gridp.N,-1))

#Analytic solution
HD = Grid.p.ymax/delta0
zDa = np.linspace(0,HD,num=1000)
##############################################################################
#Analytic solutions
##############################################################################
class analytical:
    #dimensionless fluid overpressure head
    hD = lambda zD,HD: zD + (np.exp(-HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp( zD) + \
                            (np.exp( HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp(-zD)
                            
    #dimensionless relative volumetric flux of fluid w.r.t. solid velocity: qD = - d hD/ dzD 
    qD = lambda zD,HD: -1 - (np.exp(-HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp( zD) + \
                            (np.exp( HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp(-zD)
                            
    #dimensionless fluid pressure pD = hD - zD
    pD = lambda zD,HD: (np.exp(-HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp( zD) + \
                       (np.exp( HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp(-zD)
    
    #dimensionless solid velocity potential, setting c4 = 0
    uD = lambda zD,HD: -zD - (np.exp(-HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp( zD) - \
                             (np.exp( HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp(-zD)
    
    #dimensionless solid velocity vD = - d uD/d zD 
    vD = lambda zD,HD:  1  + (np.exp(-HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp( zD) - \
                             (np.exp( HD) - 1)/(np.exp(HD) - np.exp(-HD))* np.exp(-zD)

#Initial condition
mu = mu_max  * np.ones_like(Yc_col)  #Constt viscosity
phi= phi_min * np.ones((Grid.p.N,1)) #+ (phi_max - phi_min)*(Yc(:)/Grid.p.ymax);  #Constt porosity or Decays with depth


#simulation name
simulation_type = 'inst_comp_column' #'lid_driven_cavity_flow_with_no_slip'   #lid_driven_cavity_flow_with_slip or 'no_flow' 
simulation_name = f'{simulation_type}_domain{Gridp.xmax-Gridp.xmin}by{Gridp.ymax-Gridp.ymin}_N{Gridp.Nx}by{Gridp.Ny}'

#building operators
D, Edot, Dp, Gp, I, Ms, Mp = build_stokes_ops_Darcy_Stokes(Grid)
Mud = spdiags((Ms @ (mu*(1-phi))).T,0,np.shape(Edot)[0],np.shape(Edot)[0])
Zd = build_Zd(G,phi,m,mu,Grid.p)
Kd = build_Kd(k0,n,phi,mu_f,Grid.p,Mp)

A  = 2.0 * D @ Mud @ Edot
L  = bmat([[A + Gp @ Zd @ Dp, -(Delta_rho * grav)* Gp], [Dp, -(Delta_rho * grav)* Dp @ Kd @ Gp]],format="csr")
#fs = csr_matrix((Grid.N, 1), dtype=np.float64)
fs = build_RHS(phi,Kd,Grid.p,Mp,Dp,rho_f,rho_s,Gamma,grav)


#Boundary conditions
if 'lid_driven_cavity_flow_with_slip' in simulation_type:
    BC.dof_dir =  np.concatenate((Grid.dof_pene, \
                                  Grid.dof_ymax_vt,\
                                  Grid.dof_pc))
    BC.g = np.transpose([np.concatenate((np.zeros(Grid.N_pene), \
                                         vt*np.ones(len(Grid.dof_ymax_vt)),\
                                         [0.0] ))])
elif 'inst_comp_column' in simulation_type:
    BC.dof_dir =  np.concatenate((Grid.dof_pene, \
                                  Grid.dof_ymax_vt[1:-1],\
                                  Grid.dof_ymin_vt[1:-1],\
                                  np.array([Grid.p.Nf+1])))
    BC.g = np.transpose([np.concatenate((np.zeros(Grid.N_pene), \
                                         np.zeros(len(Grid.dof_ymax_vt[1:-1])),\
                                         np.zeros(len(Grid.dof_ymin_vt[1:-1])),\
                                         np.array([analytical.hD(Grid.p.dy/(2*delta0),HD)*delta0])))])
    
elif 'lid_driven_cavity_flow_with_no_slip' in simulation_type: #did not work
    BC.dof_dir =  np.concatenate((Grid.dof_pene, \
                                  Grid.dof_ymax_vt[1:-1],\
                                  Grid.dof_ymin_vt[1:-1],\
                                  Grid.dof_xmax_vt[1:-1],\
                                  Grid.dof_xmin_vt[1:-1],\
                                  Grid.dof_pc-1))
    BC.g = np.transpose([np.concatenate((np.zeros(Grid.N_pene), \
                                         vt*np.ones(len(Grid.dof_ymax_vt[1:-1])),\
                                         np.zeros(len(Grid.dof_ymin_vt[1:-1])),\
                                         np.zeros(len(Grid.dof_xmax_vt[1:-1])),\
                                         np.zeros(len(Grid.dof_xmin_vt[1:-1])),\
                                        [0.0]))])

else:#no flow
    BC.dof_dir =  np.concatenate((Grid.dof_solid_bnd, \
                                  Grid.dof_pc))
    BC.g        = np.concatenate((np.zeros(Grid.N_solid_bnd), \
                                  [0.0]))  

BC.dof_neu = np.array([])
[B,N,fn] = build_bnd(BC,Grid,I)

#Solving for Stokes flow
u = solve_lbvp(L,fs+fn,B,BC.g,N)
v = u[:Grid.p.Nf,:]; h = u[Grid.p.Nf:,:]       #Extracting solid velocity and fluid pressure
[PSI,psi_min,psi_max] = comp_streamfun(v,Gridp)#Solid velocity stream function

p = Delta_rho * grav * (h - Yc_col)           #Overpressure [Pa] 
pf= p - rho_s * grav * Yc_col                 #Fluid pressure [Pa]
ps= pf- G * mu/ phi**m * (Dp @ v)             #Solid pressure [Pa]

#Fluid velocity
#vf = v - spdiags((1./(Mp @ phi)).T,0,Grid.p.Nf,Grid.p.Nf) @ Kd @ (Gp @ p + rho_f * grav * np.vstack([np.zeros((Grid.p.Nfx,1)),np.ones((Grid.p.Nfy,1))]))
vf = v - spdiags((1./(Mp @ phi)).T,0,Grid.p.Nf,Grid.p.Nf) @ Kd @ (Gp @ h)

[PSIf,psif_min,psif_max] = comp_streamfun(vf,Gridp) #Fluid velocity stream function

#Plotting
fig, (ax1, ax2, ax3) = plt.subplots(1,3, figsize=(15,7))
fig.suptitle(f'Dim-less depth = {HD}, Porosity = {phic}')

ax1.plot(h[0:Grid.p.Ny],Grid.p.yc,'k.',label=r'Numerical')
ax1.plot(analytical.hD(zDa,HD)*delta0,zDa*delta0,'r-',label=r'Analytical')
ax1.set_ylabel('$z$ [m]')
ax1.set_xlabel('$h$ [m]')
ax1.legend(loc = 'best')

ax2.plot(p[0:Grid.p.Ny],Grid.p.yc,'k.',label=r'N')
ax2.plot(analytical.pD(zDa,HD)*Delta_rho*grav*delta0,zDa*delta0,'r-',label=r'A')
ax2.set_xlabel('Overpressure [Pa]')

ax3.plot(v[Grid.p.Nfx:Grid.p.Nfx+Grid.p.Ny],Grid.p.yc,'k.',label=r'N')
ax3.plot(analytical.vD(zDa,HD)*Kc,zDa*delta0,'r-',label=r'A')
ax3.set_xlabel('Solid velocity [m/s]')

plt.tight_layout()
plt.savefig(f"instant_comp_column_phic{phic}_HD{HD}.pdf")

File: HW11-2DDarcy-Stokes-Steady/MATLAB/darcy_stokes_validation_compaction_colum.m
%Darcy-Stokes at a steady state (h-form)
%Author: Mohammad Afzal Shadab
%Date: April 29, 2024

set(groot,'defaultAxesFontName','Times')
set(groot,'defaultAxesFontSize',20)
set(groot,'defaulttextinterpreter','latex')
set(groot,'defaultAxesTickLabelInterpreter','latex')
set(groot,'defaultLegendInterpreter','latex')
set(groot, 'DefaultFigureVisible', 'on');

%Parameters
mu_max = 1e14; %Maximum solid viscosity [Pa.s]
phi_min= 1e-3; phi_max= 0.7; %Minimum and Maximum solid porosities [-]
G      = 1.0; %Coefficient in the bulk viscosity formulation [-]
k0     = 5.6e-11; %Relative permeability [m^2] {Meyer and Hewitt, 2017}
mu_f   = 1e-3;%Viscosity of the fluid [Pa.s]
m      = 1;   %Power law coefficient in compaction viscosity = (G/phi^m * mu_s) [-]
n      = 2;   %Power law coefficient in porosity permeability relationship k = k0*phi^n [-]
rho_s  = 917; %Density of solid [kg/m^3]
rho_f  = 1e3; %Density of fluid [kg/m^3]
Gamma  = 0;   %Rate of melting [kg/m^3-s]
grav   = 9.81;   %Acceleration due to gravity []m/s^2]
Delta_rho = rho_f - rho_s; %Difference in density of the two phases [kg/m^3]

vt     = 0;   %Tangential velocity [m/s]

%Compaction length
phic = phi_min
delta0 = sqrt(k0*phic^n*mu_max/(phic^m*mu_f))
Kc = k0*Delta_rho*grav*phic^n/mu_f


%% Build staggered grids
Gridp.xmin = 0*delta0; Gridp.xmax = 1*delta0; Gridp.Nx = 5;
Gridp.ymin = 0*delta0; Gridp.ymax = 1*delta0; Gridp.Ny = 100;
Grid = build_stokes_grid(Gridp);
[Xc,Yc] = meshgrid(Grid.p.xc,Grid.p.yc);

%mu = mu_max*(Yc(:)/Grid.p.ymax).^n;  %linearly decaying viscosity with depth


%Analytic solution
HD = Grid.p.ymax/delta0;
zDa = linspace(0,HD,1e3);
% coefficients
c1 = @(H) (exp(-H)-1)./(exp(H)-exp(-H));
c2 = @(H) (exp(H)-1)./(exp(H)-exp(-H));
% potentials
hDa = @(z,H) z + c1(H).*exp(z) + c2(H).*exp(-z);
uDa = @(z,H) -z - c1(H).*exp(z) - c2(H).*exp(-z);
% overpressure
pDa = @(z,H) c1(H).*exp(z) + c2(H).*exp(-z);
% flux & velocity
qDa = @(z,H) -1 - c1(H).*exp(z) + c2(H).*exp(-z);
vDa = @(z,H) 1 + c1(H).*exp(z) - c2(H).*exp(-z);


%Initial condition
mu = mu_max*ones(size(Yc(:)));  %This will be a function of temperature later but is constant right now
phi= phi_min * ones(Grid.p.N,1);%+ (phi_max - phi_min)*(Yc(:)/Grid.p.ymax);  %Decays with depth

%% Build Stokes operators
[D,Edot,Dp,Gp,Z,I,Ms,Mp] = build_stokes_ops_Darcy_Stokes(Grid);

%Evaluating different means
%Mud = comp_mean(mu,Ms,-1,Grid.p,1); %Average viscosity
Mud =  spdiags(Ms * (mu .* (1-phi)), 0, length(Edot),length(Edot));
Zd  = build_Zd(G,phi,m,mu,Grid.p);
Kd  = build_Kd(k0,n,phi,mu_f,Grid.p,Mp); 

A = D*2*Mud*Edot;
L = [A + Gp*Zd*Dp,  -Delta_rho*grav*Gp;...
     Dp          ,  -Delta_rho*grav*Dp*Kd*Gp ];
%fs = spalloc(Grid.N,1,0);
fs = build_RHS(phi,Kd,Grid.p,Mp,Dp,rho_f,rho_s,Gamma,grav);

%% Build BC's
BC.dof_dir = [Grid.dof_ymax_vt(2:end-1);...  % tangential velocity on the top
              Grid.dof_pene;...     % no penetration on all bnd's
              Grid.dof_ymin_vt(2:end-1);...   
              Grid.dof_pc_comp_col];         % pressure constraint
BC.dof_f_dir = [];         
BC.g       = [zeros(length(Grid.dof_ymax_vt(2:end-1)),1);...       % tangential velocity on the top
              zeros(Grid.N_pene,1);...      % no penetration on all bnd's
              zeros(length(Grid.dof_ymin_vt(2:end-1)),1);...  
              hDa(Grid.p.dy/(2*delta0),HD)*delta0];                           % pressure constraint
[B,N,fn] = build_bnd(BC,Grid,I);

%% Solve for Stokes flow
u = solve_lbvp(L,fs+fn,B,BC.g,N);
v = u(1:Grid.p.Nf); h = u(Grid.p.Nf+1:end); %Solid velocity and overpressure head
PSI = comp_streamfun(v,Grid.p);             %Solid velocity stream function

p  = (Delta_rho * grav) * (h - Yc(:));    %Overpressure [Pa]
pf =  p - rho_s * grav * Yc(:);           %Fluid pressure [Pa]
ps =  pf- G * mu./ phi.^m .* (Dp * v);      %Solid pressure
%Fluid velocity
vf = v - spdiags(1./(Mp * phi),0,Grid.p.Nf,Grid.p.Nf) * Kd * (Gp * p + rho_f * grav * [zeros(Grid.p.Nfx,1);ones(Grid.p.Nfy,1)]);
PSIf = comp_streamfun(vf,Grid.p);            %Fluid velocity stream function



hhh=figure()
set(gcf,'units','points','position',[0,0,3125,1250])
% Enlarge figure to full screen.
set(gcf, 'Position', [50 50 1500 600])
t=sgtitle(sprintf('Dim-less depth = %.1f, Porosity = %0.3f',HD, phic)); t.FontSize = 20;
%% Plotting and post-processing
subplot 131
plot(hDa(zDa,HD)*delta0,zDa*delta0,'linewidth',2), hold on
plot(h(1:Grid.p.Ny),Grid.p.yc,'--','linewidth',2)
xlabel('h [m]','fontsize',22)
ylabel('z [m]','fontsize',22)
legend('analytic','numerical','location','northwest')
set(gca,'fontsize',18)

subplot 132
plot(pDa(zDa,HD)*Delta_rho*grav*delta0,zDa*delta0,'linewidth',2), hold on
plot(p(1:Grid.p.Ny),Grid.p.yc,'--','linewidth',2)
xlabel('Overpressure [Pa]','fontsize',22)
ylabel('z [m]','fontsize',22)
legend('analytic','numerical','location','northeast')
set(gca,'fontsize',18)

subplot 133
plot(vDa(zDa,HD)*Kc,zDa*delta0,'linewidth',2), hold on
plot(v(Grid.p.Nfx+1:Grid.p.Nfx+1+Grid.p.Ny),Grid.p.yf,'--','linewidth',2)
xlabel('Solid velocity [m/s]','fontsize',22)
ylabel('z[m]','fontsize',22)
legend('analytic','numerical','location','northeast')
set(gca,'fontsize',18)
saveas(hhh,sprintf('instant_comp_column_phic%d_HD%d.png',phic,HD));


function Zd = build_Zd(G,phi,m,mu,Grid) %building zeta^*_phi at cell centers
    Zd = (G ./ (phi.^m) - 2/3) .* mu .* (1-phi);
    Zd = spdiags(Zd,0,Grid.N,Grid.N);
end

function Kd = build_Kd(k0,n,phi,mu_f,Grid,Mp) %building Kd at cell faces
    Kd = Mp * ( k0 .* phi.^n ./ mu_f);
    Kd = spdiags(Kd,0,Grid.Nf,Grid.Nf);
end

function F = build_RHS(phi,Kd,Grid,Mp,Dp,rho_f,rho_s,Gamma,grav)
    Delta_rho = rho_f - rho_s;
    %fv at cell faces
    fv = -Delta_rho*grav*(Mp*(1-phi)).* [zeros(Grid.Nfx,1); ones(Grid.Nfy,1)] ;
    
    %fp at cell centers
    fp =-(rho_f - rho_s)/(rho_f * rho_s) * Gamma * ones(Grid.N,1);
    
    F = [fv;fp];
end